src/EditorFeatures/Core/Implementation/ChangeSignature/AbstractChangeSignatureCommandHandler.cs

```diff
@@ -70,62 +71,86 @@ private bool ExecuteCommand(ITextView textView, ITextBuffer subjectBuffer, Comma
                     return false;
                 }
 
-                var reorderParametersService = document.GetLanguageService<AbstractChangeSignatureService>();
-                var result = reorderParametersService.ChangeSignature(
+                var reorderParametersService = document.GetRequiredLanguageService<AbstractChangeSignatureService>();
+                var result = reorderParametersService.ChangeSignatureAsync(
                     document,
                     caretPoint.Value.Position,
-                    (errorMessage, severity) =>
-                    {
-                        // We are about to show a modal UI dialog so we should take over the command execution
-                        // wait context. That means the command system won't attempt to show its own wait dialog 
-                        // and also will take it into consideration when measuring command handling duration.
-                        context.OperationContext.TakeOwnership();
-                        workspace.Services.GetService<INotificationService>().SendNotification(errorMessage, severity: severity);
-                    },
-                context.OperationContext.UserCancellationToken);
-
-                if (result == null || !result.Succeeded)
-                {
-                    return true;
-                }
+                    context.OperationContext.UserCancellationToken).WaitAndGetResult(context.OperationContext.UserCancellationToken);
```

Comment:
So where in this process does the options dialog get displayed? Because now that this is async, how are we getting back to the UI thread to display the dialog? Does this need a JTF-friendly wait here as opposed to WaitAndGetResult?

Reply:
@jasonmalinowski see the HandleResult method that runs after this WaitAndGetResult
https://github.com/dotnet/roslyn/pull/47768/files#diff-becc7ba48505ea7db15e854f845d4702e0d44964ec9a05d20b65c26572049596R86
The service just returns the failure kind and doesn't do any UI thread operations, so WaitAndGetResult is fine.  Then since we're already on the UI thread, HandleResult can just display the notification without any switching.

Reply:
And to add to that - the preview window is also part of HandleResult.  If the result has preview changes then it will use the preview service to display them from the UI thread.